pkg/apis/scheduling/types.go

```diff
@@ -23,9 +23,17 @@ const (
 	// that do not specify any priority class and there is no priority class
 	// marked as default.
 	DefaultPriorityWhenNoDefaultClassExists = 0
```

Comment:
Having it not part of any priority class object does not sound right to me. why to treat lowest or default value differently than others?

Reply:
0 isn't the lowest... you can have negative values. 0 is just the default.

Reply:
yeah sure, I did not mean that,  I just meant to have a class associated with each priority class value.

Reply:
when a pod does not specify a priorityClass, and there are no priority class objects, defaulting to 0 makes sense to me.

Reply:
Conceptually i think it'd be better to default to a `default priority class` (with whatever value to it) not to some integer value.

Reply:
> Conceptually i think it'd be better to default to a default priority class (with whatever value to it) not to some integer value.

We do, if one exists. If not, we have to pick a value, and since 0 is the value all previously existing pods would have, it makes sense to use here as well.

Reply:
Actually I am aware of that. if we have to pick 0 when default class does not exist, why not then just create an immortal class for that always like these system classes? the whole logic would still work the same way. Atleast we would always be dealing with classes, not this special case.

Reply:
Defaulting to zero happens in API defaulting. We don't need (or have access to in conversion) a priority class to represent that.

Reply:
I am not seeing if the value is being set to default to 0 at API level, at least thats what i thought https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/scheduling/v1alpha1/zz_generated.defaults.go .
But this is being used here and set to default in admission plugin if we did set to default in the above we would not need this i think : https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/priority/admission.go#L243 .

Reply:
ok i see the value is int32 so should not get 0. then i think we dont need this: https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/priority/admission.go#L243 and this special variable DefaultPriorityWhenNoDefaultClassExists. right?

Reply:
Ignore the above. let me check again.

Reply:
I checked priority class, whereas i should have checked pod's priority value. let me check that.

Reply:
Anyway I am ok with defaulting at API level but then I am wondering why we need DefaultPriorityWhenNoDefaultClassExists?

Reply:
I am assuming when you say its getting default to 0, its somewhere in https://github.com/kubernetes/api/blob/master/core/v1/generated.pb.go based on https://github.com/kubernetes/api/blob/master/core/v1/generated.proto#L3192. But as I said, in that case, I dont think we need any new constant like this `DefaultPriorityWhenNoDefaultClassExists`.

Reply:
Priority field is an *int32, not int32. `DefaultPriorityWhenNoDefaultClassExists` is needed to set the default.

Reply:
I am not sure how else to convince that if we are not defaulting to 0 at API level, it would be better to default based on a priority class to keep it in sync with other cases where priority of a pod is pulled from a priority class. Again if we are defaulting to 0 at API level that is different thing and thats perfectly fine, but then we would not need DefaultPriorityWhenNoDefaultClassExists.

Reply:
@aveshagarwal 
Two points:
1. The concern of this PR is something else. It would be better to have a separate issue to discuss such matters. I dislike the fact that we mix discussions of unrelated matters and slow down the flow of development. I have heard the same complaints from other colleagues as well.
2. We should be backward compatible and support Pods that do not have any priority class. So, we must not assume that all Pods have a priority class.
3. Priority resolution happens in admission controller. So does priority resolution for pods without any PriorityClass.